cloud_function_go/function.go

```diff
@@ -0,0 +1,159 @@
+// Package p contains an HTTP Cloud Function.
+package p
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+const maxSize = 10485760
+
+type logzioConfig struct {
+	token      string
+	listener   string
+	typeLog    string
+	validation bool
+}
+
+func shouldRetry(statusCode int) bool {
+	retry := true
+	switch statusCode {
+	case http.StatusBadRequest:
+		fmt.Printf("Got HTTP %d bad request, skip retry\n", statusCode)
+		retry = false
+	case http.StatusNotFound:
+		fmt.Printf("Got HTTP %d not found, skip retry\n", statusCode)
+		retry = false
+	case http.StatusUnauthorized:
+		fmt.Printf("Got HTTP %d unauthorized, skip retry. Please check if you providing proper token or listner\n", statusCode)
+		retry = false
+	case http.StatusForbidden:
+		fmt.Printf("Got HTTP %d forbidden, skip retry\n", statusCode)
+		retry = false
+	case http.StatusOK:
+		retry = false
+	}
+	return retry
+}
+
+func (l *logzioConfig) validateAndPopulateArguments(r *http.Request) {
+
+	token := r.URL.Query().Get("token")
+	if len(l.token) == 0 {
+		fmt.Printf("Logzio token must be provided")
+		l.validation = false
+	} else {
+		l.token = token
+	}
+
+	listener := r.URL.Query().Get("listener")
+	if len(l.listener) == 0 {
+		fmt.Printf("Logzio listener must be provided")
+		l.validation = false
+	} else {
+		l.listener = listener
+	}
+
+	typeLog := r.URL.Query().Get("type")
+	if len(l.typeLog) == 0 {
+		fmt.Printf("Set default log type, `pubsub`")
+		l.typeLog = "pubsub"
+	} else {
+		l.typeLog = typeLog
+	}
+
+}
+
+func doRequest(rawDecodedText []byte, url string) {
+	// gzip compress data before shipping
+	var compressedBuf bytes.Buffer
+	gzipWriter := gzip.NewWriter(&compressedBuf)
+	_, err := gzipWriter.Write(rawDecodedText)
+	if err != nil {
+		fmt.Printf("Failed to compress log")
+		return
+	}
+	err = gzipWriter.Close()
+	if err != nil {
+		fmt.Printf("Failed to close gzip")
+		return
+	}
+	if binary.Size(compressedBuf) > maxSize {
```

Comment:
If it bigger we should cut it
Also the size checking should be before trying to compress the buffer